certbot-ci/certbot_integration_tests/certbot_tests/test_main.py

```diff
@@ -92,9 +90,154 @@ def test_renew_with_hook_scripts(context):
 
     assert_cert_count_for_lineage(context.config_dir, certname, 1)
 
+    misc.generate_test_file_hooks(context.config_dir, context.hook_probe)
+    context.certbot(['renew'])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    assert_hook_execution(context.hook_probe, 'deploy')
+
+
+def test_renew_files_propagate_permissions(context):
+    """Test proper certificate renewal with custom permissions propagated on private key."""
+    certname = context.get_domain('renew')
+    context.certbot(['-d', certname])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 1)
+
+    os.chmod(join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)), 0o444)
+    context.certbot(['renew'])
+
+    assert_cert_count_for_lineage(context.config_dir, certname, 2)
+    assert_equals_permissions(
+        join(context.config_dir, 'archive/{0}/privkey1.pem'.format(certname)),
+        join(context.config_dir, 'archive/{0}/privkey2.pem'.format(certname)), 0o074)
```

Comment:
It's super unlikely that this would ever be a problem, but can we also add a check that privkey2 is world readable? This check exists in the shell script at https://github.com/certbot/certbot/blob/master/tests/certbot-boulder-integration.sh#L360 and it's possible that Certbot modified `privkey1.pem` so the two keys compare equal, but world permission wasn't properly preserved.

Reply:
I added the assertion.